pkg/cmd/cli/cmd/rsync/rsync.go

```diff
@@ -229,7 +236,89 @@ func (o *RsyncOptions) Validate() error {
 
 // RunRsync copies files from source to destination
 func (o *RsyncOptions) RunRsync() error {
-	return o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	if err := o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut); err != nil {
+		return err
+	}
+
+	if !o.Watch {
+		return nil
+	}
+	return o.WatchAndSync()
+}
+
+// WatchAndSync sets up a recursive filesystem watch on the sync path
+// and invokes rsync each time the path changes
+func (o RsyncOptions) WatchAndSync() error {
+
+	// these variables must be accessed while holding the changeLock
+	// mutex as they are shared between goroutines to communicate
+	// sync state/events.
+	var (
+		changeLock sync.Mutex
+		dirty      bool
+		lastChange time.Time
+		watchError error
+	)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error setting up filesystem watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				changeLock.Lock()
+				glog.V(5).Infof("filesystem watch event: %s", event)
+				lastChange = time.Now()
+				dirty = true
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					if e := watcher.Remove(event.Name); e != nil {
+						glog.Warningf("error removing watch for %s: %v", event.Name, e)
+					}
+				} else {
+					watchError = fsnotification.AddRecursiveWatch(watcher, event.Name)
+				}
+				changeLock.Unlock()
+			case err := <-watcher.Errors:
+				changeLock.Lock()
+				watchError = fmt.Errorf("error watching filesystem for changes: %v", err)
+				changeLock.Unlock()
+			}
+		}
+	}()
+
+	err = fsnotification.AddRecursiveWatch(watcher, o.Source.Path)
+	if err != nil {
+		// If this fails, it may be necessary to increase available watches on linux
+		// by running 'sudo sysctl fs.inotify.max_user_watches=65536'
+		return fmt.Errorf("error watching source path %s: %v", o.Source.Path, err)
+	}
+
+	for {
+		changeLock.Lock()
+		if watchError != nil {
+			return watchError
+		}
+		// if a change happened more than 2 seconds ago, sync it now.
+		// if a change happened less than 2 seconds ago, sleep for 2 seconds
+		// and see if more changes happen, we don't want to sync when
+		// the filesystem is in the middle of changing due to a massive
+		// set of changes (such as a local build in progress).
+		if dirty && time.Now().After(lastChange.Add(2*time.Second)) {
```

Comment:
The two occurrences of `2 * time.Second` are interrelated, how about it becoming a named variable?
Then the comments can refer to the variable instead of a particular duration (remember the code/comment pair we were fixing some time earlier in extended tests, I think, where we got them out of sync?).